cmd/crypto/vault.go

```diff
@@ -130,79 +130,70 @@ func NewVault(config VaultConfig) (KMS, error) {
 	return v, nil
 }
 
-// renewSecret tries to renew the given secret. It blocks
-// until it receives either the new secret or encounters an error.
-func (v *vaultService) renewSecret(secret *vault.Secret) (*vault.Secret, error) {
+// reauthenticate() tries to login in 1 minute
+// intervals until successful.
+func (v *vaultService) reauthenticate() {
+	retryDelay := 1 * time.Minute
+	go func() {
+		for {
+			if err := v.authenticate(); err != nil {
+				time.Sleep(retryDelay)
+				continue
+			}
+			return
+		}
+	}()
+}
+
+// renewer calls vault client's renewer that automatically
+// renews secret periodically
+func (v *vaultService) renewer(secret *vault.Secret) {
 	renewer, err := v.client.NewRenewer(&vault.RenewerInput{
 		Secret: secret,
 	})
 	if err != nil {
-		logger.CriticalIf(context.Background(), fmt.Errorf("crypto: failed to create hashicorp vault renewer: %s", err))
+		logger.FatalIf(err, "crypto: hashicorp vault token renewer could not be started")
 	}
+	v.tokenRenewer = renewer
 	go renewer.Renew()
 	defer renewer.Stop()
 
 	for {
 		select {
 		case err := <-renewer.DoneCh():
 			if err != nil {
-				return nil, err
+				v.reauthenticate()
+				renewer.Stop()
```

Comment:
We already do `defer` so why need to do explicit `renewer.Stop()` ?

Reply:
unnecessary, so removed it.